### PR TITLE
Manual Decaf

### PR DESCRIPTION
**Pull request by confused-Techie**
EDIT::

@DeeDeeG  @Sertonix this PR is now ready for review. I've gone ahead and decaffed all the files here except `scope-selector-parser.js` since that isn't actually decaf work, it is created from `scope-selector-parser.pegjs` and it seems out of scope for this PR to attempt at making that code pretty.

But otherwise all comments have been added in, and a major aspect is I've tried to walk the line from decaf to refactor on this code. So in some locations there may be better ways to write the code or to handle some logic, but I've instead opted to have the most honest translation between CoffeeScript and JavaScript instead. At a later time a refactor will likely be a great idea, but at least for the time being we know it works, and know what's trying to be accomplished.

Any kind of review would be very appreciated, so that I can move onto the next steps of trying to remove Oniguruma in it's Native Module form from this repo and get us one step closer to Electron 19.

Thanks for any comments!

---

This PR starts the process of decaffeinating `first-mate`.

As discussed on our [Discussions](https://github.com/orgs/pulsar-edit/discussions/70) before this PR was started, a branch of `first-mate` was created called `decaf`. This `decaf` branch used only the built in `grunt` tools to get our JS source, with **zero** other modifications. Meaning it is exactly what we have been using in production.

Then from there on `decafe-manual-ct` the work of manually decaffeinating the source files was started.

This PR does a few major things, changing the test files to `require` the JavaScript source instead of the CoffeeScript source, so we can ensure our tests actually run, and are running successfully.

Then this PR also changing `test` in our `package.json` to instead of using `grunt` to run tests (because `grunt` would decafe all over again, overwriting our changes, to instead run `jasmine-focused` within our script to test directly against the JavaScript files.

Then finally the work of actually decaffeinating itself began. But I know there was some debate about doing all of the files at once, or some of them. So I tried to find a happy middle, working on three files for now, and keeping each file as it's own PR, so that each file itself is easy to review.

But onto the actual coding done. 

Each file has had any comments that were in it's source CoffeeScript added back in, as well as classes were rewritten to have modern JavaScript syntax patterns. Then of course removing needless returns, and exporting our functions in the normal sane way.

If this looks good, then this work can continue, otherwise if it's easy I can begin to start work on the other files as well. Or once merged into `decaf` others can continue this work on the other remaining files.

Once work has been completed, we can remove `prepublish` and `grunt` scripts, as well as all CoffeeScript related dependencies.

@DeeDeeG I'd appreciate your two cents on this format.

---

EDIT: I've also updated the GH Actions test runner, to skip any decaf work, and to run tests on Pull Requests, so we can properly test these PR's as they are made.

---

DOUBLE EDIT:

Seems actions just weren't enabled for this repo. So whoops, but in that and testing, also removed the `prepublish` script from our `package.json` to prevent it attempting to run when installing the package.

### CODE REVIEW

**Comment by confused-Techie** [comment]
Gotta be honest, not sure why the tests aren't running on this PR. But the action is valid yaml syntax, and seems fine to me. I can attest to all tests passing locally, but strange the tests aren't running here.

File: lib/first-mate.js
Comment (sertonix) [inline lib/first-mate.js]: You can also use a newer style for this:
```js
module.exports = {
  ScopeSelector: require('./scope-selector'),
  GrammarRegistry: require('./grammar-registry'),
  Grammar: require('./grammar'),
  get OnigRegExp(){
    return require('oniguruma').OnigRegExp;
  }
};
```
Otherwise this pull request looks good!

Comment (confused-Techie) [reply lib/first-mate.js]: Great catch there, I'll go ahead and add it in. But thanks for the review and if it's looking good I'll keep going in this style and hopefully is anyone else working on decaf work

File: lib/grammar-registry.js
Comment (sertonix) [inline lib/grammar-registry.js]: This should be the same as:
```js
return new Disposable(() => 
  this.removeGrammar(grammar)
);
```
But maybe I should just make a pull request with my changes ideas later.

Comment (confused-Techie) [reply lib/grammar-registry.js]: You totally could, but I'll test to see if that works. Thanks!

**Comment by confused-Techie** [comment]
This PR is now ready for review @Sertonix

**Comment by DeeDeeG** [comment]
Sorry for the delay, I got sidetracked on another thing for the ppm repo.

_(I delayed commenting on this so I could get a chance to look over the JS code and make any finer comments I could come up with, as I expected the thing for ppm to take me only a day, but it took 3 days instead. Oops.)_

I wanted to say from the first day this was posted that I agree with the philosophy of the general approach... in no small part because it's just like the approach I proposed for the ppm repo, and started at that repo's `decaf-central` branch.

I like the clean, logically separated commits at the beginning of this branch! Very good for review purposes.

I can see that the diff between the original, machine transpilation, and the final result after manual changes, is pretty huge, so it ends up being not so much easier to review anyhow. But it's better than nothing! (Whenever I see a file that was fully deleted and replaced with a fully new file, 100% red/`-` diff of the old file and 100% green/`+` diff of the new file, it just makes reasoning about those files so much harder for me, personally speaking.) At least now each chunk that changed can be compared right next to the original chunk. There is some contextualization of the individual changes this way. So it may not be totally easy to review, but it's easier than just "old file deleted, new file added."

I leave the finer points of the JS review to others who are more fluent in JS and comfortable writing it -- I am shy about reviewing others' JS code myself. (It does pass the existing specs in CI! I like to see that!) So for those doing such review, I hope this approach has made things easier for you. If it has, then I hope that proves this approach is a success and can be generalized for decaf at the other repos.

File: package.json
Comment (DeeDeeG) [inline package.json]: I think any module that is an explicitly written dependency in `package.json` can be run by name from the `package.json` scripts.

So I think this can just be:
```suggestion
    "test": "jasmine-focused --coffee --captureExceptions spec",
```

~~And we have to remember to make sure to compile the `.pegjs` files again, I assume, since it looks like Grunt had been handling that before during the `grunt prepublish` steps. See `Gruntfile.coffee` for the details of that.~~ Ah, okay. So the pegJS is already transpiled and committed here as just `.js` alongside the decaf stuff. I see. Safe to disregard this crossed out part of this comment, then.

Comment (confused-Techie) [reply package.json]: Great catch there. Yeah they should be able to run like you suggested. And might as well do some cleanup, and add grunt (or another method) to compile the `pegjs` files so we can continue using normal scripts there.

But would you be on board with compiling the `pegjs` via GitHub Action's, rather than a prepublish script. Since I've love to be able to cut down on what we have to do when building the editor itself

Comment (DeeDeeG) [reply package.json]: Well the version of `first-mate` package on the npm registry has the PEGjs compiled already. `lib/scope-selector-parser.js`. (Because the `prepublish` script runs before the tarball is packed and uploaded to npmjs.com.)

Whatever we need to do to have it as plain JS for consumption by the Pulsar repo/the Pulsar editor is pretty much okay by me.

Assuming we do want to use this forked version straight off of GitHub.com repo URLs in Pulsar's `package.json`, then it would be fine to have it done as a GitHub action... then pushed to some `dist` branch??? Or transpiled once and just committed to the main branch as plain JS.

I just thought of something... Any URLs pointing to a package tarball will be treated pretty much equally by npm. So if we want to package our tarballs using `npm pack` in GitHub actions and post that packed tarball as a GitHub Releases artifact... (this all could be automated as a "publish release" workflow, I think) we can then use the direct download URL to that tarball from the GitHub Releaes in Pulsar's `package.json` and it should work fine. Just a thought.

**Comment by confused-Techie** [comment]
@DeeDeeG I appreciate the review. And hope it makes it slightly easier, even if they are basically still huge rewrites. For me at least it is easier.

Since here you just compare ugly, overly generic JS with specified JS, whereas against straight CoffeeScript, you have to transpile in your head, then go from whatever your in your head JS looked like to what was written.

And no worries on taking your time with the review, theres a lot going on here always so take the time needed.

As for your other comments on ensuring we are compiling the PegJS code, I mentioned I would like to see that done as a GitHub Action, to offload the responsibility on use and CI systems once we are actually building the editor (Because we all know there's to much going on when building anyway). But to do this I'd like to add a GitHub Action that runs `pegjs` as a dependency. Thing is, we use `grunt-peg` here to compile. So really like I mentioned as soon as this is merged I have another PR ready to go with improvements specifically relating to our use of a native oniguruma that I'd like to remove. And I can add those changes in there, since they feel out of scope on this PR, which the decaf branch already has a compiled pegjs, so the code will work without this action, at least until we start changing `scope-selector-parser.js`, so I feel it's safe to wait until the next PR on this issue. But feel free to let me know what you think, and hopefully we can get this one merged

**Review by DeeDeeG — APPROVED** [review]
"Rubber stamp" approve of these actual changes/diff/code since I trust you folks on the JS stuff. And the tests are passing without modifications really to speak of.

---

For the "meta" conversation about "how to decaf":

I did realize something about the big diff, which reminded me of @GuilleW's "PR 2" suggestion. First and foremost the big diff seems to be coming from the "remove wrapper" commit -- from taking out `(function() {main code here}).call(this)` wrapping all the code.

So circling back to @GuilleW's suggestion about removing the wrapper as its own PR, I realize how that would move that large (but not super important) diff out of the way. So the "finishing touches" manual decaf PR would have a clearer and smaller diff.

(Since you have write access to the repos, I don't think you need to do the wrapper removal as a PR, just commit it to the decaf base branch. It makes some sense for someone without push rights to the repository to suggest doing it as a PR, just as a practical means to get the code to the repo somehow. But in all reality, when done properly, the "PR 1" and "PR 2" from @GuilleW's plan, as actual PRs, would not invite much commentary other than a quick "looks good thanks!" and merge.)

(Removing the wrapper and extracting the file contents out one level implies an indentation change, which makes the whole file appear in the diff, minus some double line breaks that "didn't change". So it's the whole file as a diff sort of arbitrarily broken up by the double newlines (paragraph breaks basically) that the diff algorithm decides were common code unchanged from before/after the wrapper removal. But in effect the diff is the entire file.)

Maybe it doesn't matter much -- since I am deferring to other people to review the JS, whatever those people find comfortable is good enough. I would be more inclined to attempt to review the thoughtful manual changes if the diff was more targeted and smaller, but no guarantees, I suppose, since I might still defer to other people more comfortable with reviewing JS.

---

P.S. thanks for doing this, because it tests out the decaf process and teaches us things about how that works in practice.

**Comment by confused-Techie** [comment]
@DeeDeeG appreciate your rubber stamp lol.

But great points about this. But in the future we could defer to those with write access to do the initial setup if we'd like. Something I'd be willing to do with a ping before someone starts decaf work on a repo.

Now would we rather go that method then for the fix, of removing the biggest wrapper? Even if that means the current resulting code wouldn't function.

I'm all for it, but my only concern would be the inability for those working from the decaf branch to test properly, since everything would fail tests at first and as the decaf goes would be a countdown really in failing tests since there are none. Which at least for me, I relied on all green tests throughout the decaf process to ensure my changes were true to the original behavior.

I do agree that doing this would make it easier to review, but think it might lean to fair on making the reviewers life easier while making the contributors life harder.

No for something like this, we could essentially have someone do the initial decaf work. The big non-logic changes. Such as removing the wrapping and throwing everything into a class that actually makes sense. But that does seem to get kinda close to being actual decaf work.

So sorry kinda rambling, but I'd love it if we are able to remove the wrapper without causing tests to start failing. But otherwise thanks for the approval I'll go ahead and merge this one, because I am quite excited to see if I can remove another native module

**Comment by DeeDeeG** [comment]
> I'd love it if we are able to remove the wrapper without causing tests to start failing.

Hmm. I just assumed removing the wrapper doesn't change anything functionally. I personally agree about wanting to avoid publishing any commits that break tests if it's avoidable.

I've been diving into this decaf process myself a tad, using my local clone of the ppm repo, for learning purposes. So I'm learning a bit.

If anyone can give me pointers about why removing the wrappers breaks the code or changes its functionality, I'd be curious. But only if anyone feels up to it.

Tangent: `coffeescript@^v2.7.0` module can decaf the ppm repo, with passing tests. So we might wanna use `coffee` compiler v2.x there, so long as we're confident in the code it produces? Hmm.

(Apparently the breaking changes from CoffeeScript 1.x to 2.x are meant to be limited to some uncommon/edge case scenarios that won't necessarily occur in most CoffeeScript 1.x code. Per: https://coffeescript.org/#breaking-changes)

**Comment by confused-Techie** [comment]
@DeeDeeG from what I recall seeing,

I believe that the wrapper lets the code be self callable. Since it's not a standard object in it's current form there's nothing to 'call' it. The wrapper lets code call itself in an async way, and the wrapper used is commonly what you will find in the startup or init scripts of Node Applications from CLI.

**Comment by GuilleW** [comment]
A wrapper make you sure every variables inside can't be change by another script. Main usage is for web page. 

**Example:** If you include  your JS file in html page, every variables declared in your file will be global: you can't predict what happen with other scripts, framework, etc (collision, bugs...).

In our case, in Pulsar, WE know what we're doing since the JS file is not intended to be include in another way. It's safe enough to remove wrapper. We can let wrapper, but it's just more code for no really good advantage.

**Pull request merged**
